Rework OSDK intros + Golang examples (#28372)
openshift/openshift-docs: modules/osdk-golang-implement-controller.adoc

// Module included in the following assemblies:
//
// * operators/operator_sdk/osdk-golang-tutorial.adoc

[id="osdk-golang-implement-controller_{context}"]
= Implementing the controller

After creating a new API and controller, you can implement the controller logic.

.Procedure

* For this example, replace the generated controller file `controllers/memcached_controller.go` with the `memcached_controller.go` link:https://github.com/operator-framework/operator-sdk/blob/v1.2.0/testdata/go/memcached-operator/controllers/memcached_controller.go[example implementation].
+

> The example controller executes the following reconciliation logic for each `Memcached` custom resource (CR):

Reviewer: > ! execute verb
Use only if a simpler term, such as "run", "issue", "start", or "enter" is inappropriate.

+
--
* Create a Memcached deployment if it does not exist.
* Ensure that the deployment size is the same as specified by the `Memcached` CR spec.
* Update the `Memcached` CR status with the names of the `memcached` pods.
--

> The next subsections explain how the controller watches resources and how the reconcile loop is triggered. You can skip these subsections to go directly to building and running the Operator.

Reviewer: Same as the previous module - you can't mix procedure and concept data.

[id="osdk-golang-implement-controller-resources_{context}"]
== Resources watched by the controller

The `SetupWithManager()` function in `controllers/memcached_controller.go` specifies how the controller is built to watch a CR and other resources that are owned and managed by that controller.

[source,go]
----
import (
	...
	appsv1 "k8s.io/api/apps/v1"
	...
)

func (r *MemcachedReconciler) SetupWithManager(mgr ctrl.Manager) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(&cachev1alpha1.Memcached{}).
		Owns(&appsv1.Deployment{}).
		Complete(r)
}
----

`NewControllerManagedBy()` provides a controller builder that allows various controller configurations.

> `For(&cachev1alpha1.Memcached{})` specifies the `Memcached` type as the primary resource to watch. For each `Memcached` type Add, Update, or Delete event, the reconcile loop is sent a reconcile `Request` argument, a namespace and name key, for that `Memcached` object.

Reviewer: The second sentence is missing something. PTAL?

Author: Awkward sentence casualty of trying to remove a parentheses, IIRC.

`Owns(&appsv1.Deployment{})` specifies the `Deployment` type as the secondary resource to watch. For each `Deployment` type Add, Update, or Delete event, the event handler maps each event to a reconcile request for the owner of the deployment. In this case, the owner is the `Memcached` object for which the deployment was created.

[id="osdk-golang-implement-controller-configs_{context}"]
== Controller configurations

> There are many other useful configurations that can be made when initializing a controller. For example:

Reviewer: > Similarly, avoid using expletive constructions such as it is, there are, and there is, which hide the subject of the sentence.

I'd also make this sentence active.

> * Set the max number of concurrent reconciles for the controller using the `MaxConcurrentReconciles` option, which defaults to `1`:

Reviewer: ```suggestion
* Set the maximum number of concurrent reconciles for the controller by using the `MaxConcurrentReconciles` option, which defaults to `1`:
```

+
[source,go]
----
func (r *MemcachedReconciler) SetupWithManager(mgr ctrl.Manager) error {
    return ctrl.NewControllerManagedBy(mgr).
        For(&cachev1alpha1.Memcached{}).
        Owns(&appsv1.Deployment{}).
        WithOptions(controller.Options{
            MaxConcurrentReconciles: 2,
        }).
        Complete(r)
}
----

> * Filter watch events using predicates.

Reviewer: Is your user going to know what a predicate is?

> * Choose the type of link:https://godoc.org/sigs.k8s.io/controller-runtime/pkg/handler#hdr-EventHandlers[EventHandler] to change how a watch event translates to reconcile requests for the reconcile loop. For Operator relationships that are more complex than primary and secondary resources, the `EnqueueRequestsFromMapFunc` handler can be used to transform a watch event into an arbitrary set of reconcile requests.

Reviewer: ```suggestion
* Choose the type of link:https://godoc.org/sigs.k8s.io/controller-runtime/pkg/handler#hdr-EventHandlers[EventHandler] to change how a watch event translates to reconcile requests for the reconcile loop. For Operator relationships that are more complex than primary and secondary resources, you can use the `EnqueueRequestsFromMapFunc` handler to transform a watch event into an arbitrary set of reconcile requests.
```

----

[NOTE]
====
You can return `Result` with `RequeueAfter` set to periodically reconcile a CR.
====

For more on reconcilers, clients, and interacting with resource events, see the link:https://sdk.operatorframework.io/docs/building-operators/golang/references/client/[Controller Runtime Client API] documentation.

[id="osdk-golang-implement-controller-rbac-markers_{context}"]
== Permissions and RBAC manifests

The controller requires certain RBAC permissions to interact with the resources it manages. These are specified using RBAC markers, such as the following:

[source,go]
----
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds/status,verbs=get;update;patch
// +kubebuilder:rbac:groups=cache.example.com,resources=memcacheds/finalizers,verbs=update
// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;

func (r *MemcachedReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
  ...
}
----

> The `ClusterRole` object manifest at `config/rbac/role.yaml` is generated from the above markers using the `controller-gen` utility whenever the `make manifests` command is run.

Reviewer: > ! above adjective, adverb, preposition
1. Do not use to refer to a subsequent software version or fix level. Use "later". 2. Do not use to indicate a relative location in a document, as in "the above restrictions" or "the steps above". Use "previous" or "preceding". Examples: "The previous restrictions apply to ..."; "The preceding steps..."
```suggestion
The `ClusterRole` object manifest at `config/rbac/role.yaml` is generated from the previous markers by using the `controller-gen` utility whenever the `make manifests` command is run.
```
